File: hash/swift/RosettaImageHash/Sources/RosettaImageHash/Internal/Haar.swift
import Foundation

/// 2-D Haar DWT/IDWT with pywt's 'symmetric' boundary mode.
///
/// CRITICAL: SQRT2_INV = (0.5).squareRoot() (NOT 1.0 / 2.0.squareRoot()).
/// The latter is one ULP lower (0.7071067811865475 vs 0.7071067811865476)
/// and accumulates errors through 8-10 wavedec levels, flipping bits at
/// the whash median boundary. Confirmed across Java/Go/Rust/JS ports.
///
/// dwt2 evaluates column-pass BEFORE row-pass (required for float addition order).
/// idwt2 reverses axis order: row-pass first, then column-pass.

private let SQRT2_INV: Double = (0.5).squareRoot()

struct DWT2Result {
    let cA: [[Double]]
    let cH: [[Double]]
    let cV: [[Double]]
    let cD: [[Double]]
}

struct WavedecResult {
    var cA: [[Double]]
    /// Each entry is [cH, cV, cD] for one level. Outer-to-inner = deepest-first.
    var details: [[[[Double]]]]
}

func dwt2(_ x: [[Double]]) -> DWT2Result {
    let h = x.count
    let w = x[0].count

    // Column pass first
    var colLow = [[Double]](repeating: [], count: w)
    var colHigh = [[Double]](repeating: [], count: w)
    for xCol in 0..<w {
        var col = [Double](repeating: 0, count: h)
        for y in 0..<h { col[y] = x[y][xCol] }
        let (low, high) = dwt1d(col)
        colLow[xCol] = low
        colHigh[xCol] = high
    }
    let outH = (h + 1) >> 1
    let outW = (w + 1) >> 1

    var cA = [[Double]](repeating: [Double](repeating: 0, count: outW), count: outH)
    var cH = [[Double]](repeating: [Double](repeating: 0, count: outW), count: outH)
    var cV = [[Double]](repeating: [Double](repeating: 0, count: outW), count: outH)
    var cD = [[Double]](repeating: [Double](repeating: 0, count: outW), count: outH)

    for y in 0..<outH {
        var rowLow = [Double](repeating: 0, count: w)
        var rowHigh = [Double](repeating: 0, count: w)
        for xCol in 0..<w {
            rowLow[xCol] = colLow[xCol][y]
            rowHigh[xCol] = colHigh[xCol][y]
        }
        let (lowL, highL) = dwt1d(rowLow)
        let (lowH, highH) = dwt1d(rowHigh)
        for xd in 0..<outW {
            cA[y][xd] = lowL[xd]
            cV[y][xd] = highL[xd]
            cH[y][xd] = lowH[xd]
            cD[y][xd] = highH[xd]
        }
    }
    return DWT2Result(cA: cA, cH: cH, cV: cV, cD: cD)
}

func idwt2(cA: [[Double]], cH: [[Double]], cV: [[Double]], cD: [[Double]]) -> [[Double]] {
    let sh = cA.count
    let sw = cA[0].count
    let outH = sh * 2

    var colLow = [[Double]](repeating: [], count: sw)
    var colHigh = [[Double]](repeating: [], count: sw)
    for xCol in 0..<sw {
        var ll = [Double](repeating: 0, count: sh)
        var lh = [Double](repeating: 0, count: sh)
        var hl = [Double](repeating: 0, count: sh)
        var hh = [Double](repeating: 0, count: sh)
        for y in 0..<sh {
            ll[y] = cA[y][xCol]
            lh[y] = cV[y][xCol]
            hl[y] = cH[y][xCol]
            hh[y] = cD[y][xCol]
        }
        colLow[xCol] = idwt1d(low: ll, high: lh)
        colHigh[xCol] = idwt1d(low: hl, high: hh)
    }

    var out = [[Double]](repeating: [], count: outH)
    for y in 0..<outH {
        var low = [Double](repeating: 0, count: sw)
        var high = [Double](repeating: 0, count: sw)
        for xCol in 0..<sw {
            low[xCol] = colLow[xCol][y]
            high[xCol] = colHigh[xCol][y]
        }
        out[y] = idwt1d(low: low, high: high)
    }
    return out
}

func wavedec2(_ x: [[Double]], level: Int) -> WavedecResult {
    var current = x
    var details: [[[[Double]]]] = []
    for _ in 0..<level {
        let r = dwt2(current)
        details.append([r.cH, r.cV, r.cD])
        current = r.cA
    }
    details.reverse()
    return WavedecResult(cA: current, details: details)
}

func waverec2(_ d: WavedecResult) -> [[Double]] {
    var current = d.cA
    for level in d.details {
        current = idwt2(cA: current, cH: level[0], cV: level[1], cD: level[2])
    }
    return current
}

private func dwt1d(_ x: [Double]) -> (low: [Double], high: [Double]) {
    var xx = x
    var n = xx.count
    if (n & 1) != 0 {
        xx.append(xx[n - 1])
        n += 1
    }
    let half = n >> 1
    var low = [Double](repeating: 0, count: half)
    var high = [Double](repeating: 0, count: half)
    for i in 0..<half {
        let a = xx[2 * i]
        let b = xx[2 * i + 1]
        low[i] = SQRT2_INV * a + SQRT2_INV * b
        high[i] = SQRT2_INV * a - SQRT2_INV * b
    }
    return (low, high)
}

private func idwt1d(low: [Double], high: [Double]) -> [Double] {
    let half = low.count
    let n = half * 2
    var out = [Double](repeating: 0, count: n)
    for i in 0..<half {
        out[2 * i] = SQRT2_INV * low[i] + SQRT2_INV * high[i]
        out[2 * i + 1] = SQRT2_INV * low[i] - SQRT2_INV * high[i]
    }
    return out
}

File: hash/js/rosetta-squint-hash/src/internal/haar.ts
/**
 * 2-D Haar DWT/IDWT with pywt's 'symmetric' boundary mode.
 *
 * IMPORTANT: SQRT2_INV = Math.sqrt(0.5) (NOT 1 / Math.sqrt(2)).
 * The latter is one ULP lower and accumulates errors through 8-10 wavedec
 * levels, flipping bits at the whash median boundary. Confirmed across
 * Java/Go/Rust ports.
 *
 * Column-pass before row-pass evaluation order matches pywt's float
 * addition order. Required for byte-exact parity.
 */

const SQRT2_INV = Math.sqrt(0.5); // 0.7071067811865476 — DO NOT use 1 / Math.sqrt(2)

export interface Dwt2Result {
  cA: number[][];
  cH: number[][];
  cV: number[][];
  cD: number[][];
}

export interface WavedecResult {
  cA: number[][];
  /** Each entry is [cH, cV, cD] for one level. Outer-to-inner = deepest-first. */
  details: number[][][][];
}

export function dwt2(x: number[][]): Dwt2Result {
  const h = x.length;
  const w = x[0].length;

  // Column pass first
  const colLow: number[][] = new Array(w);
  const colHigh: number[][] = new Array(w);
  const col = new Array<number>(h);
  for (let xCol = 0; xCol < w; xCol++) {
    for (let y = 0; y < h; y++) col[y] = x[y][xCol];
    const [low, high] = dwt1d(col);
    colLow[xCol] = low;
    colHigh[xCol] = high;
  }
  const outH = (h + 1) >> 1;
  const outW = (w + 1) >> 1;

  const cA: number[][] = [];
  const cH: number[][] = [];
  const cV: number[][] = [];
  const cD: number[][] = [];
  for (let y = 0; y < outH; y++) {
    cA.push(new Array<number>(outW).fill(0));
    cH.push(new Array<number>(outW).fill(0));
    cV.push(new Array<number>(outW).fill(0));
    cD.push(new Array<number>(outW).fill(0));
  }
  const rowLow = new Array<number>(w);
  const rowHigh = new Array<number>(w);
  for (let y = 0; y < outH; y++) {
    for (let xCol = 0; xCol < w; xCol++) {
      rowLow[xCol] = colLow[xCol][y];
      rowHigh[xCol] = colHigh[xCol][y];
    }
    const [lowL, highL] = dwt1d(rowLow);
    const [lowH, highH] = dwt1d(rowHigh);
    for (let xd = 0; xd < outW; xd++) {
      cA[y][xd] = lowL[xd];
      cV[y][xd] = highL[xd];
      cH[y][xd] = lowH[xd];
      cD[y][xd] = highH[xd];
    }
  }
  return { cA, cH, cV, cD };
}

export function idwt2(
  cA: number[][],
  cH: number[][],
  cV: number[][],
  cD: number[][],
): number[][] {
  const sh = cA.length;
  const sw = cA[0].length;
  const outW = sw * 2;

  // Row-pass first (inverse of col-pass-first in dwt2):
  //   cA and cV share the col-low-pass origin -> combine along rows
  //   cH and cD share the col-high-pass origin -> combine along rows
  const rowLow: number[][] = [];
  const rowHigh: number[][] = [];
  for (let y = 0; y < sh; y++) {
    rowLow.push(idwt1d(cA[y], cV[y]));
    rowHigh.push(idwt1d(cH[y], cD[y]));
  }

  // Column-pass second: for each column, combine rowLow col with rowHigh col
  const out: number[][] = [];
  for (let y2 = 0; y2 < sh * 2; y2++) {
    out.push(new Array<number>(outW).fill(0));
  }
  for (let xCol = 0; xCol < outW; xCol++) {
    const low = new Array<number>(sh);
    const high = new Array<number>(sh);
    for (let y = 0; y < sh; y++) {
      low[y] = rowLow[y][xCol];
      high[y] = rowHigh[y][xCol];
    }
    const col = idwt1d(low, high);
    for (let y2 = 0; y2 < sh * 2; y2++) {
      out[y2][xCol] = col[y2];
    }
  }
  return out;
}

export function wavedec2(x: number[][], level: number): WavedecResult {
  let current: number[][] = x.map(row => row.slice());
  const details: number[][][][] = [];
  for (let l = 0; l < level; l++) {
    const { cA, cH, cV, cD } = dwt2(current);
    details.push([cH, cV, cD]);
    current = cA;
  }
  details.reverse();
  return { cA: current, details };
}

export function waverec2(d: WavedecResult): number[][] {
  let current = d.cA;
  for (const level of d.details) {
    current = idwt2(current, level[0], level[1], level[2]);
  }
  return current;
}

function dwt1d(x: number[]): [number[], number[]] {
  let n = x.length;
  let xx = x;
  if ((n & 1) !== 0) {
    const ext = x.slice();
    ext.push(x[n - 1]);
    xx = ext;
    n = n + 1;
  }
  const half = n >> 1;
  const low = new Array<number>(half);
  const high = new Array<number>(half);
  for (let i = 0; i < half; i++) {
    const a = xx[2 * i];
    const b = xx[2 * i + 1];
    low[i] = SQRT2_INV * a + SQRT2_INV * b;
    high[i] = SQRT2_INV * a - SQRT2_INV * b;
  }
  return [low, high];
}

function idwt1d(low: number[], high: number[]): number[] {
  const half = low.length;
  const n = half * 2;
  const out = new Array<number>(n);
  for (let i = 0; i < half; i++) {
    out[2 * i] = SQRT2_INV * low[i] + SQRT2_INV * high[i];
    out[2 * i + 1] = SQRT2_INV * low[i] - SQRT2_INV * high[i];
  }
  return out;
}
